spx-backend/internal/controller/controller.go

```diff
@@ -497,28 +498,40 @@ func (ctrl *Controller) FmtCode(ctx context.Context, input *FmtCodeInput) (*fmtc
 type UpInfo struct {
 	// Uptoken
 	Token string `json:"token"`
-	// Valid time for uptoken, unit: second
-	Expires uint64 `json:"expires"`
+	// Timestamp after which the uptoken is considered expired
+	ExpiresAt uint64 `json:"expiresAt"`
+	// Maximum file size allowed
+	MaxSize int64 `json:"maxSize"`
 	// Bucket Region
 	Region string `json:"region"`
 	// Base URL to fetch file
 	BaseUrl string `json:"baseUrl"`
 }
 
 func (ctrl *Controller) GetUpInfo(ctx context.Context) (*UpInfo, error) {
-	var expires uint64 = 1800 // second
 	putPolicy := qiniuStorage.PutPolicy{
 		Scope:        ctrl.kodo.bucket,
+		Expires:      1800, // 30 minutes
 		ForceSaveKey: true,
 		SaveKey:      "files/$(etag)/$(fname)",
-		Expires:      expires,
+
+		// There is a chance that object key collision happens, so we
+		// use insertOnly mode to prevent unintended overwrites.
+		InsertOnly: 1,
+
+		// The hardcoded size limit here should be sufficient for most
+		// frontend use cases. If needed, we can make it configurable in
+		// the future.
+		FsizeLimit: 25 << 20, // 25 MiB
 	}
 	mac := qiniuAuth.New(ctrl.kodo.ak, ctrl.kodo.sk)
 	upToken := putPolicy.UploadToken(mac)
+	expiresAt := uint64(time.Now().Unix()) + putPolicy.Expires
 	return &UpInfo{
-		Token:   upToken,
-		Expires: expires,
```

Comment:
https://github.com/goplus/builder/issues/487 这里提到的

> This will allow clients to know the exact expiration time, which is more informative than just knowing the duration in seconds.

其中“informative”对应的具体的好处是什么样的？

之前用 expires 是考虑客户端的时间可能不正确；假设客户端时间比服务端快了 30 分钟，那么刚获取到的 token 也会被认为过期了，客户端会永远无法得到它认为可用的 token；用 expires 可以消除这个影响。

expires 也有自己的误差：从服务端生成 token 到客户端代码得到到 token 会有时间差，不过这个时间差可以认为是可控的：客户端跟服务端都会有超时的逻辑；前端代码里原来的

```ts
bufferTime = 5 * 60 * 1000 // refresh uptoken 5min before it expires
```

是预期来消除这个误差的影响。而如果客户端时间不正确，它跟服务端时间的误差就不太可控了

Reply:
> 之前用 expires 是考虑客户端的时间可能不正确；假设客户端时间比服务端快了 30 分钟，那么刚获取到的 token 也会被认为过期了，客户端会永远无法得到它认为可用的 token；用 expires 可以消除这个影响。

嗯这个考虑很合理，那我改回去。

Reply:
done 1926898